src/lib.rs

```diff
@@ -4442,6 +4443,90 @@ pub fn digraph_spiral_layout(
     layout::spiral_layout(&graph.graph, scale, center, resolution, equidistant)
 }
 
+fn _num_shortest_paths_unweighted<Ty: EdgeType>(
+    graph: &StableGraph<PyObject, PyObject, Ty>,
+    source: usize,
+) -> PyResult<HashMap<usize, u128>> {
```

Comment:
I am concerned about overflows here. The number of paths can grow very fast, it may not fit on a 128-bit integer. Can you try using [BigInt](https://docs.rs/num-bigint/0.4.0/num_bigint/struct.BigInt.html) and see how much it affects the performance? [PyO3 has supports it](https://github.com/PyO3/pyo3/blob/05e4bdf69cde2bc49865fbafe756a8507dd45f1d/src/num_bigint.rs#L39-L48).

Reply:
> Commenting because I am a bit concerned about using 128-bit integers for a combinatorics. Counting functions can grow very fast, and if there is a lot of symmetry then they grow even faster.
> 
> We are adding this function to help some folks speed up their code, but I don't want to fall into the overflow benchmark trap. It is useless to have a speed up if the result is not correct because of an overflow

Sure that's a good point, while the [max value of u128](https://doc.rust-lang.org/std/primitive.u128.html#associatedconstant.MAX) is: 340282366920938463463374607431768211455 which I honestly can't see being an issue for most people on this in practice; it could come up with large graphs. I did think that rust would panic on overflow for us, and it sort of does. By default only in debug mode addition panics on overflow, but in release mode (which is what we obviously ship in precompiled binaries on pypi) it wraps around on overflow which we definitely don't want (I mistakenly thought it always would panic). I'll benchmark both this current PR as it is, changing the addition to explicitly use a function call with overflow checking, and using `BigInt` and we can go from there.

Reply:
I ran some quick benchmarks on this and it's 4-6x slower than using u128. But in that testing and benchmarking your point about the rapid growth of the count here was more obvious. The largest square grid graph I could run on without overflowing a u128 was 66x66. At a 66x66 graph the u128 took 0.0001586 sec and the BigUint version took 0.0004135 sec. So I'm comfortable with a 4x slowdown if we're that fast since anything slower would likely mean an invalid result or an overflow exception.

Reply:
For a graph with `n` nodes, the number of shortest paths between two fixed nodes is bounded by the number of simple paths between these nodes, which is bounded by `n - 2`. So you will not get overflow using `u128`.